### PR TITLE
CFE-214 : SC-6 : CaC rule that checks workloads for resource requests/limits

### PR DESCRIPTION
**Pull request by Connect2naga**
CFE-214 : Inital commit with basic rule to validate resource request

#### Description:

  When deploying an application, it is important to tune based on memory and CPU consumption,
  allocating enough resources for the application to function properly. Images provided by
  OpenShift Dedicated behave properly within the confines of the memory they are allocated.
  However, any applicaiton images must pay attention to the specific resources required to
  ensure they are available.

  If the node where a Pod is running has enough of a resource available, it's possible (and allowed)
  for a container to use more resource than its request for that resource specifies.
  However, a container is not allowed to use more than its resource limit.

#### Rationale:

  Resource requests/limits provide constraints that limit aggregate resource consumption
  per container. This helps prevent resource starvation. When deploying your
  application, it is important to tune based on memory and CPU consumption,
  allocating enough resources for the application to function properly.

- Fixes # CFE-214

### CODE REVIEW

File: applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/tests/ocp4/e2e.yml
Comment (Connect2naga) [inline applications/openshift/api-server/api_server_no_adm_ctrl_plugins_disabled/tests/ocp4/e2e.yml:2]: Changed to check other failures.

File: applications/openshift/general/resource_requests_limits_in_daemonset/rule.yml
Comment (Vincent056) [inline applications/openshift/general/resource_requests_limits_in_daemonset/rule.yml:36]: I think we should skip openshift-* and kube-* namespaces

Comment (Connect2naga) [reply applications/openshift/general/resource_requests_limits_in_daemonset/rule.yml:36]: Done.
initially i am also thinking same, there is possibility that openshift-* & kube-* resource should cause problem(i mean use more resource if no limits).

**Comment by lmzuccarelli** [comment]
@Connect2naga - just a question would it not have been easier to have checked the Quality Of Service status (Best Effort) if this set then it means that the request/limits where not set (Burstable and Guaranteed are when they are set) ?

File: applications/openshift/general/resource_requests_limits_in_daemonset/test/has_resource_requests_Limits.pass.sh
Comment (lmzuccarelli) [inline applications/openshift/general/resource_requests_limits_in_daemonset/test/has_resource_requests_Limits.pass.sh:262]: @Connect2naga - correct me if I'm wrong should this just skip as the namespace starts with 'kube-' i.e kube-system) even if we have resource/limits set ?

Comment (Connect2naga) [reply applications/openshift/general/resource_requests_limits_in_daemonset/test/has_resource_requests_Limits.pass.sh:262]: yes Luigi, it will skip these two namespaces.

Comment (Connect2naga) [reply applications/openshift/general/resource_requests_limits_in_daemonset/test/has_resource_requests_Limits.pass.sh:262]: yes this scenario's we will cover in unit test.

Comment (Connect2naga) [reply applications/openshift/general/resource_requests_limits_in_daemonset/test/has_resource_requests_Limits.pass.sh]: added unit test case for namespace skip

**Comment by lmzuccarelli** [comment]
overall lgtm. I will leave the final review and approval to the compliance team

**Comment by Connect2naga** [comment]
> overall lgtm. I will leave the final review and approval to the compliance team

Thanks @lmzuccarelli  for review.

**Review by Vincent056 — APPROVED** [review]
Thanks for the PR, I think they look good!

**Comment by jhrozek** [comment]
Looking quite good. I have some questions before merging:
 - can you please run `git rebase origin/master` and push to your fork's branch again to get rid of the merge commits?
 - can you remove the managed fields from the unit tests to make the YAML manifests more compact and better readable?
 - is there a reason to check separately for DSs/deployments/statefulsets but not for pods? What about e.g. cronjobs or in general pods scheduled and managed by operators/apps w/o supervision?

**Comment by Connect2naga** [comment]
@jhrozek  thanks for review comments,
Point 1 & 2 taken,
Point 3 : we can directly scan the pods instead of separately, but we have to suggest how to avoid so we need to provide the originator(deployment....),  so i feel we can find out one level above & not started with pods..
is it better approach scanning the pods? Need your suggestion, how we handled with other rules in similar case?

**Comment by jhrozek** [comment]
> @jhrozek thanks for review comments, Point 1 & 2 taken, 

thanks, I didn't have time to debug the failing CI checks, but I'll look at them tomorrow

> Point 3 : we can directly scan the pods instead of separately, but we have to suggest how to avoid so we need to provide the originator(deployment....), so i feel we can find out one level above & not started with pods.. is it better approach scanning the pods? Need your suggestion, how we handled with other rules in similar case?

You know, you are right. After some discussion on slack I think your approach is good, the best defense would be a webhook that rejects workloads without requests or limits, the compliance checks are just FYI and to illustrate compliance, not a defense measure.

So I think this will be good to go once CI passes.

**Comment by Connect2naga** [comment]
Thanks @jhrozek

**Comment by Connect2naga** [comment]
HI @jhrozek ,
Could you please check the CI Failure , Because these test case not related to correct change.
suspecting some dependency issue.

**Comment by jhrozek** [comment]
Sorry it took me so long. This seems to be the issue in the CIS profile at least:
```
     helpers.go:805: Result - Name: e2e-cis-controller-insecure-port-disabled - Status: FAIL - Severity: low
    helpers.go:808: E2E-FAILURE: The expected result for the controller_insecure_port_disabled rule didn't match. Expected 'PASS', Got 'FAIL' 
```

In PCI-DSS I see:
```
     helpers.go:805: Result - Name: e2e-pci-dss-configure-network-policies-namespaces - Status: FAIL - Severity: high
    helpers.go:808: E2E-FAILURE: The expected result for the configure_network_policies_namespaces rule didn't match. Expected 'PASS', Got 'FAIL'
    helpers.go:805: Result - Name: e2e-pci-dss-controller-insecure-port-disabled - Status: FAIL - Severity: low
    helpers.go:808: E2E-FAILURE: The expected result for the controller_insecure_port_disabled rule didn't match. Expected 'PASS', Got 'FAIL' 
```
did this PR touch those test files or the rules? (Very short on time today, sorry, just have enough time to check the CI logs..)

**Comment by Vincent056** [comment]
> Sorry it took me so long. This seems to be the issue in the CIS profile at least:
> 
> ```
>      helpers.go:805: Result - Name: e2e-cis-controller-insecure-port-disabled - Status: FAIL - Severity: low
>     helpers.go:808: E2E-FAILURE: The expected result for the controller_insecure_port_disabled rule didn't match. Expected 'PASS', Got 'FAIL' 
> ```
> 
> In PCI-DSS I see:
> 
> ```
>      helpers.go:805: Result - Name: e2e-pci-dss-configure-network-policies-namespaces - Status: FAIL - Severity: high
>     helpers.go:808: E2E-FAILURE: The expected result for the configure_network_policies_namespaces rule didn't match. Expected 'PASS', Got 'FAIL'
>     helpers.go:805: Result - Name: e2e-pci-dss-controller-insecure-port-disabled - Status: FAIL - Severity: low
>     helpers.go:808: E2E-FAILURE: The expected result for the controller_insecure_port_disabled rule didn't match. Expected 'PASS', Got 'FAIL' 
> ```
> 
> did this PR touch those test files or the rules? (Very short on time today, sorry, just have enough time to check the CI logs..)

Just checkout the artifact and rule `controller_insecure_port_disabled`, this fail is caused by `port\":[\"0\"]` missing from kube-controller-manager configmap, it is an independent issue with CI. @jhrozek @Connect2naga

**Comment by Connect2naga** [comment]
@Vincent056  Thanks for the analysis.
So then do we need to bring the port in`kube-controller-manager configmap`?

**Comment by Connect2naga** [comment]
@Vincent056 @Jakuje  , i tried to squash and Pick, some thing went wrong :( , Now i am seeing total 188 commit.
instead of spending time Created new PR with my changes on latest Master., 
Please review the changes : https://github.com/ComplianceAsCode/content/pull/8095

**Review by Vincent056 — APPROVED** [review]
/lgtm

**Comment by jhrozek** [comment]
There's still this:
```
 helpers.go:805: Result - Name: e2e-cis-controller-insecure-port-disabled - Status: FAIL - Severity: low
    helpers.go:808: E2E-FAILURE: The expected result for the controller_insecure_port_disabled rule didn't match. Expected 'PASS', Got 'FAIL'
```
checking why..

**Comment by jhrozek** [comment]
Ah, OK, the failures are not your fault, it's just that the port argument that rule checks for got deprecated in new k8s versions, so we need to adjust that check.

**Comment by jhrozek** [comment]
This was the commit that removed `--port`, we should instead check if `secure-port` is NOT 0 on 4.10+

**Comment by Connect2naga** [comment]
> This was the commit that removed `--port`, we should instead check if `secure-port` is NOT 0 on 4.10+: [kubernetes/kubernetes@e481d99](https://github.com/kubernetes/kubernetes/commit/e481d9996591331d6173e4cb146afcaf5e0308f1)

So `controller_insecure_port_disabled`  rule need to change then.

**Comment by jhrozek** [comment]
> > This was the commit that removed `--port`, we should instead check if `secure-port` is NOT 0 on 4.10+: [kubernetes/kubernetes@e481d99](https://github.com/kubernetes/kubernetes/commit/e481d9996591331d6173e4cb146afcaf5e0308f1)
> 
> So `controller_insecure_port_disabled` rule need to change then.

Yes, I'm working on it.

**Comment by jhrozek** [comment]
Let's see if PR #8099 helps.

**Review by Vincent056 — APPROVED** [review]

**Comment by Vincent056** [comment]
@Connect2naga the unit test still have errors:

```
INFO - Script skip_openshift_kube_ns_resource_requests_Limits.pass.sh using profile (all) OK
ERROR - Failed to execute 'cd /root/ssgts/resource_requests_limits_in_deployment; SHARED=/root/ssgts/shared bash -x no_resource_request_limits.fail.sh' on root@localhost: 'ssh' command returned non-zero.
ERROR - Terminating due to error: Failed to execute 'cd /root/ssgts/resource_requests_limits_in_deployment; SHARED=/root/ssgts/shared bash -x no_resource_request_limits.fail.sh' on root@localhost.
---------Ansible Remediation Logs---------
INFO - xccdf_org.ssgproject.content_rule_resource_requests_limits_in_deployment
INFO - Script skip_openshift_kube_ns_resource_requests_Limits.pass.sh using profile (all) OK
ERROR - Failed to execute 'cd /root/ssgts/resource_requests_limits_in_deployment; SHARED=/root/ssgts/shared bash -x no_resource_request_limits.fail.sh' on root@localhost: 'ssh' command returned non-zero.
ERROR - Terminating due to error: Failed to execute 'cd /root/ssgts/resource_requests_limits_in_deployment; SHARED=/root/ssgts/shared bash -x no_resource_request_limits.fail.sh' on root@localhost.
```

File: applications/openshift/general/resource_requests_limits_in_statefulset/tests/has_resource_requests_Limits.pass.sh
Comment (Vincent056) [inline applications/openshift/general/resource_requests_limits_in_statefulset/tests/has_resource_requests_Limits.pass.sh:140]: missed a space here, the jq filter needs to be consistent with the one in the rule.

File: applications/openshift/general/resource_requests_limits_in_statefulset/tests/skip_openshift_kubesystem_ns_resource_requests_Limits.pass.sh
Comment (Vincent056) [inline applications/openshift/general/resource_requests_limits_in_statefulset/tests/skip_openshift_kubesystem_ns_resource_requests_Limits.pass.sh:131]: here too, missing a space

File: applications/openshift/general/resource_requests_limits_in_deployment/tests/no_resource_request_limits.fail.sh
Comment (Vincent056) [inline applications/openshift/general/resource_requests_limits_in_deployment/tests/no_resource_request_limits.fail.sh:114]: here missing the space

**Review by Vincent056 — CHANGES_REQUESTED** [review]
The jq filter has to be exact the same as the one used in the rule

**Review by Vincent056 — APPROVED** [review]

**Comment by Vincent056** [comment]
I will merge the PR once it passes CI, it seems like there are some issues with cluster provision

**Comment by Connect2naga** [comment]
> I will merge the PR once it passes CI, it seems like there are some issues with cluster provision

Thanks @Vincent056 ,
 i will all so try to re-run test after some time.

**Pull request merged**
